janus-idp/backstage-showcase
# provide kubernetesAuthApi in frontend kubernetes dynamic plugin

## Description

The Kubernetes plugin has a default API factory that work right out the box with a few oidcProviders as defined [here](https://github.com/backstage/backstage/blob/d2af76ffb6a105dfec1310733f5ca8e316872116/plugins/kubernetes/src/plugin.ts#L66-L96), but `oidc.oidc` is not included. We are able to override this definition by adding our own implementation. This PR adds an additional utility API to the `backstage-plugin-kubernetes` plugin to allow the use of `oidcTokenProvider: 'oidc'` ([More on `oidcTokenProvider`](https://backstage.io/docs/features/kubernetes/configuration/#clustersoidctokenprovider-optional)).

## Which issue(s) does this PR fix

- Fixes [RHDHBUGS-36](https://issues.redhat.com/browse/RHDHBUGS-36)

## PR acceptance criteria
The error does not appear when configuring OIDC as the auth provider with the Kubernetes plugin like: 
```
kubernetes:
  ...
  clusterLocatorMethods:
    - type: 'config'
      clusters:
          ...
          authProvider: 'oidc'
          oidcTokenProvider: 'oidc'
```
[Note](https://backstage.io/docs/features/kubernetes/configuration/#clustersoidctokenprovider-optional) on `oidcTokenProvider`

![image](https://github.com/user-attachments/assets/fc3f04b9-00a9-4bb8-aa2f-ea37c41b0904)

Please make sure that the following steps are complete:


## How to test changes / Special notes to the reviewer
1. Set up oidc authentication in `auth.providers.oidc`
2. Add annotation to a component to [surface the Kubernetes components as part of an entity](https://backstage.io/docs/features/kubernetes/configuration/#surfacing-your-kubernetes-components-as-part-of-an-entity)
3. Configure the Kubernetes plugins with the following auth and confirm that no error occurs.
``` 
         authProvider: 'oidc'
         oidcTokenProvider: 'oidc'
```
